crates/turborepo-lib/src/graph/walker.rs

```diff
@@ -0,0 +1,257 @@
+use std::{collections::HashMap, hash::Hash};
+
+use futures::future::join_all;
+use petgraph::{
+    visit::{IntoNeighborsDirected, IntoNodeIdentifiers},
+    Direction,
+};
+use tokio::{
+    sync::{broadcast, mpsc, oneshot, watch},
+    task::JoinHandle,
+};
+use tracing::log::trace;
+
+pub struct Walker<N> {
+    cancel: watch::Sender<bool>,
+    node_events: Option<mpsc::Receiver<(N, oneshot::Sender<()>)>>,
+    join_handles: Vec<JoinHandle<()>>,
+}
+
+pub type WalkMessage<N> = (N, oneshot::Sender<()>);
+
+// These constraint might look very stiff, but since all of the petgraph graph
+// types use integers as node ids and GraphBase already constraints these types
+// to Copy + Eq so adding Hash + Send + 'static as bounds aren't unreasonable.
+impl<N: Eq + Hash + Copy + Send + 'static> Walker<N> {
+    /// Create a new graph walker for a DAG that emits nodes at a given
+    /// The graph should not be modified after a walker is created as the nodes
+    /// emitted might no longer exist or might miss newly added edges.
+    pub fn new<G: IntoNodeIdentifiers<NodeId = N> + IntoNeighborsDirected>(graph: G) -> Self {
+        let (cancel, cancel_rx) = watch::channel(false);
+        let mut txs = HashMap::new();
+        let mut rxs = HashMap::new();
+        for node in graph.node_identifiers() {
+            // Each node can finish at most once so we set the capacity to 1
+            let (tx, rx) = broadcast::channel::<()>(1);
+            txs.insert(node, tx);
+            rxs.insert(node, rx);
+        }
+        // We will be emitting at most txs.len() nodes so emitting a node should never
+        // block
+        let (node_tx, node_rx) = mpsc::channel(txs.len());
+        let mut join_handles = vec![];
```

Comment:
I think FuturesUnordered would probably be more efficient and I don't think we need to be strict about the order they are resolved in since we are using tasks and they will be running anyways regardless of polling or not.